application/usecases/instagram_posts.go

```diff
@@ -0,0 +1,164 @@
+	"bytes"
+	"fmt"
+	"image"
+	"image/png"
+	"net/http"
+	"os"
+	"reflect"
+
+	"github.com/Davincible/goinsta/v3"
+)
+
+type InstagramPostsController struct {
+	feedItems []FeedItem
+	codes     CodesSet
+}
+
+type FeedItem struct {
+	text  string
+	image *[]byte
+}
+
+// Use a set to avoid duplicate codes
+type CodesSet map[string]void
+
+type void struct{}
+
+const INSTAGRAM_CACHE_FILE = "./.cache/lastInstagramPosts.json"
+
+func NewInstagramPostsController() *InstagramPostsController {
+	var codes CodesSet
+	err := readJSONFileToStruct(&codes, INSTAGRAM_CACHE_FILE)
+	if err != nil {
+		config.Log.Info("Cached posts not found. Starting from ground...")
+	}
+
+	c := &InstagramPostsController{codes: codes}
+	return c
+}
+
+func (t *InstagramPostsController) Run() ([]models.TGMessage, error) {
+
+	var insta *goinsta.Instagram
+	insta, err := goinsta.Import("./.cache/.goinsta")
+	if err != nil {
+		fmt.Println("Cookies not found. Login in with user/password...")
+		insta = goinsta.New(os.Getenv("ig_user"), os.Getenv("ig_pass"))
+		err := insta.Login()
+		if err != nil {
+			fmt.Println(err)
+		}
+		defer func(ins *goinsta.Instagram, path string) {
+			err := ins.Export(path)
+			if err != nil {
+			}
+		}(insta, "./.cache/.goinsta")
+	}
+
+	initialCodes := t.codes
+	items, updatedCodes, err := getRecentMedia(insta, t)
+
+	// If there are no new posts then we don't send anything
+	if reflect.DeepEqual(initialCodes, updatedCodes) {
+		return []models.TGMessage{
+			{MSG: "", Media: nil, Kind: models.KindMedia},
+		}, nil
+	}
+
+	// Transform the items information to TGMessage objects
+	messages := itemsToMessages(items)
+	return messages, nil
+
+}
+
+func itemsToMessages(items []FeedItem) []models.TGMessage {
+	messages := []models.TGMessage{}
+	for _, item := range items {
+		message := models.TGMessage{MSG: item.text, Media: item.image, Kind: models.KindMedia}
+		messages = append(messages, message)
+	}
+	return messages
+}
+
+func getRecentMedia(insta *goinsta.Instagram, t *InstagramPostsController) ([]FeedItem, CodesSet, error) {
+
+	missingCodes := make(CodesSet)
+	var member void
+
+	acc := "tomorrowland"
+	profile, err := insta.VisitProfile(acc)
+	if err != nil {
+		fmt.Println("Cannot visit profile", err)
+	}
+
+	feed := profile.Feed
+	var items []FeedItem
+
+	num_retrieved_feeds := 5
+
+	for _, item := range feed.Items[0:num_retrieved_feeds] {
+		if _, ok := t.codes[item.Code]; !ok {
+			image, err := downloadFile(item.Images.Versions[0].URL)
+			if err != nil {
+				fmt.Println("Error downloading image", err)
+			}
+			cbbytes := imageToByteArray(image)
+
+			text := `<b>` + item.Caption.Text + "\n\n" + `</b>+Info: ` + fmt.Sprintf("https://www.instagram.com/p/%s/?img_index=1", item.Code)
+
+			var current_item = FeedItem{text: text, image: &cbbytes}
+
+			missingCodes[item.Code] = member
+			items = append(items, current_item)
+		}
+	}
+
+	// Merge missing codes with existing codes
+	t.codes = mergeCodes(t.codes, missingCodes)
```

Comment:
mergeCodes could be just removed just iterating over the map with key, val. (Go tip here):

```suggestion
        for key, val := range missingCodes {
            t.codes[key] = void{}
        }
```

Reply:
This doesn't seem to have exactly the same behaviour, will keep the working version for now.